From 6e1941ce4fd22c6934811c7d0cbd65c03a687c51 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:55:14 -0400
Subject: [fs] ext4: disable synchronous transaction batching if max_batch_time==0

Message-id: <1414058125-4183-161-git-send-email-lczerner@redhat.com>
Patchwork-id: 98243
O-Subject: [RHEL 7.1 PATCH v2 160/171] ext4: disable synchronous transaction batching if max_batch_time==0
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Eric Sandeen <sandeen@redhat.com>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 5dd214248f94d430d70e9230bda72f2654ac88a8

 The mount manpage says of the max_batch_time option,

  This optimization can be turned off entirely
  by setting max_batch_time to 0.

 But the code doesn't do that.  So fix the code to do
 that.

 Signed-off-by: Eric Sandeen <sandeen@redhat.com>
 Signed-off-by: Theodore Ts'o <tytso@mit.edu>
 Cc: stable@vger.kernel.org
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/super.c b/fs/ext4/super.c
index b442e4e..1b19874 100644
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -1523,8 +1523,6 @@ static int handle_mount_opt(struct super_block *sb, char *opt, int token,
    arg = JBD2_DEFAULT_MAX_COMMIT_AGE;
   sbi->s_commit_interval = HZ * arg;
  } else if (token == Opt_max_batch_time) {
-  if (arg == 0)
-   arg = EXT4_DEF_MAX_BATCH_TIME;
   sbi->s_max_batch_time = arg;
  } else if (token == Opt_min_batch_time) {
   sbi->s_min_batch_time = arg;
diff --git a/fs/jbd2/transaction.c b/fs/jbd2/transaction.c
index 322f81e..35db2ac 100644
--- a/fs/jbd2/transaction.c
+++ b/fs/jbd2/transaction.c
@@ -1586,9 +1586,12 @@ int jbd2_journal_stop(handle_t *handle)
   * to perform a synchronous write.  We do this to detect the
   * case where a single process is doing a stream of sync
   * writes.  No point in waiting for joiners in that case.
+  *
+  * Setting max_batch_time to 0 disables this completely.
   */
  pid = current->pid;
- if (handle->h_sync && journal->j_last_sync_writer != pid) {
+ if (handle->h_sync && journal->j_last_sync_writer != pid &&
+     journal->j_max_batch_time) {
   u64 commit_time, trans_time;
 
   journal->j_last_sync_writer = pid;
-- 
1.7.1